### PR TITLE
Added Terraform CI for build and integration tests

### PR DESCRIPTION
**Pull request by samiib**
Part 1 Terraform CI:
- Go Format check
- Go Vendor check
- Go Build check
- Terraform Acceptance Tests for APIC v5.2 & v6.0

### CODE REVIEW

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    runs-on: ubuntu-latest
+    concurrency:
+        group: tf-aci-ci-test-${{ matrix.apic_host.name }}
+        cancel-in-progress: false
+    strategy:
+      fail-fast: false
+      matrix:
+        apic_host:
+          - name: v5.2
+            url: 'https://173.36.219.83/'
```
Comment (samiib) [inline .github/workflows/checks.yml:43]: Once this is merged, I will create a different PR to update the hosts and add CI user/password.

File: .github/workflows/checks.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
+name: CI
+
+on:
+  pull_request:
+    paths:
+      - '.github/workflows/checks.yaml'
+      - '**.go'
```
Comment (lhercot) [inline .github/workflows/checks.yml:7]: Why limit to those paths? A change in the YAML or JSON files would force a new generation which might not get caught by this filter.

File: .github/workflows/checks.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      - uses: actions/setup-go@v5
+        with:
+          go-version-file: 'go.mod'
+      - run: chmod +x ./scripts/gofmtcheck.sh
+      - name: gofmt Check
+        run: ./scripts/gofmtcheck.sh
+      # - name: Generate provider code
+      #   run: go generate
+      # - name: Check generated code for diffs
+      #   run: git diff --exit-code
```
Comment (lhercot) [inline .github/workflows/checks.yml:28]: I would make this a separate job.

File: .github/workflows/checks.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+          terraform_version: '1.7.*'
+          terraform_wrapper: false
+      - name: Terraform Acceptance Test (APIC ${{ matrix.apic_host.name }})
+        run: go test github.com/CiscoDevNet/terraform-provider-aci/v2/internal/provider -v -cover -timeout 120m
+        env:
+          TF_ACC: '1'
+          TF_ACC_STATE_LINEAGE: '1'
+          ACI_VAL_REL_DN: false
+          ACI_USERNAME: ${{ secrets.TF_ACC_ACI_USERNAME }}
+          ACI_PASSWORD: ${{ secrets.TF_ACC_ACI_PASSWORD }}
```
Comment (lhercot) [inline .github/workflows/checks.yml:62]: My understanding is that secrets cannot be used when using PR from forks. They will use the forks value for that secret not the main repo.

File: .github/workflows/checks.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      - uses: actions/checkout@v4
+      - uses: actions/setup-go@v5
+        with:
+          go-version-file: 'go.mod'
+      - uses: hashicorp/setup-terraform@v2
+        with:
+          terraform_version: '1.7.*'
+          terraform_wrapper: false
+      - name: Terraform Acceptance Test (APIC ${{ matrix.apic_host.name }})
+        run: go test github.com/CiscoDevNet/terraform-provider-aci/v2/internal/provider -v -cover -timeout 120m
```
Comment (lhercot) [inline .github/workflows/checks.yml:56]: Can we get the coverage output saved somewhere? (not sure if we can also use codecov.io for golang)

**Review by lhercot — CHANGES_REQUESTED** [review]

File: .github/workflows/checks.yml (hunk 5 above)
Comment (samiib) [reply .github/workflows/checks.yml:56]: Good idea. I will look into it. 👍

File: .github/workflows/checks.yml (hunk 4 above)
Comment (samiib) [reply .github/workflows/checks.yml:62]: I plan to remove these secrets and reuse the hosts/username/password from the ansible-aci repo after this is merged.

File: .github/workflows/checks.yml (hunk 3 above)
Comment (samiib) [reply .github/workflows/checks.yml:28]: Agreed. When I get it working I will separate it.

File: .github/workflows/checks.yml (hunk 2 above)
Comment (samiib) [reply .github/workflows/checks.yml:7]: True, but the changed/added generated code should still be `.go`
I'll remove it anyway, I think this was from the Terraform build example.

File: .github/workflows/checks.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          - name: azure_cloud
+            url: 'https://20.245.236.136/'
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-go@v5
+        with:
+          go-version-file: 'go.mod'
+      - uses: hashicorp/setup-terraform@v2
+        with:
+          terraform_version: '1.7.*'
```
Comment (akinross) [inline .github/workflows/checks.yml:62]: We will be introducing 1.8 specific features probably, should we run it on terraform 1.8? Or fix CI to run on latest?

File: .github/workflows/checks.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+          terraform_wrapper: false
+      - name: Terraform Acceptance Test (APIC ${{ matrix.apic_host.name }})
+        run: go test github.com/CiscoDevNet/terraform-provider-aci/v2/internal/provider -v -cover -timeout 120m
```
Comment (akinross) [inline .github/workflows/checks.yml:65]: Is the timeout sufficient or should we increase to higher number?

File: .github/workflows/checks.yml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    branches: [master]
+  pull_request:
+
+permissions:
+  # Permission for checking out code
+  contents: read
+
+jobs:
+  build:
+    name: Check & Build Provider
```
Comment (akinross) [inline .github/workflows/checks.yml:14]: Should we run a `go generate` in the check and build step, to ensure there are no differences during generate?

Comment (samiib) [reply .github/workflows/checks.yml:14]: https://github.com/CiscoDevNet/terraform-provider-aci/pull/1203#discussion_r1577223549
Yes, this has already been a WIP. However, just need to fix some issues with the generator running in the CI

File: .github/workflows/checks.yml (hunk 7 above)
Comment (samiib) [reply .github/workflows/checks.yml:65]: For now, a happy test is running between 10-15 mins. Hence, I believe 2 hrs should be enough but we can always adjust if it's required.

File: .github/workflows/checks.yml (hunk 6 above)
Comment (samiib) [reply .github/workflows/checks.yml:62]: I think we should run it with whichever version we build with. I matched it with the Terraform plugin version in the go.mod file.
We just need to make sure we change both places at the same time.
I don't think it should fix to always using latest. Unpinned versions in CI can cause more issues that are hard to diagnose.

Comment (akinross) [reply .github/workflows/checks.yml:62]: ok we will likely soon need 1.8 features so then we need to adjust then. agree with comment about latest

File: .github/workflows/checks.yml (hunk 7 above)
Comment (samiib) [reply .github/workflows/checks.yml]: Changed to 5h as discussed.

File: .github/workflows/checks.yml [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      - name: gofmt Check
+        run: ./scripts/gofmtcheck.sh
+      - run: go mod tidy
+      - run: go mod vendor
+      - name: Check vendor for changes
+        run: git diff --exit-code
+      - name: Build
+        run: go build -v
+
+  # diff:
```
Comment (samiib) [inline .github/workflows/checks.yml:31]: I will implement this in a different PR.

File: .github/workflows/checks.yml [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      ACI_USERNAME: 'ansible_github_ci'
+      ACI_PASSWORD: 'sJ94G92#8dq2hx*K4qh'
+    concurrency:
+        group: tf-aci-ci-test-${{ matrix.apic_host.name }}
+        cancel-in-progress: false
+    strategy:
+      fail-fast: false
+      matrix:
+        apic_host:
+          # - name: v4.2
```
Comment (samiib) [inline .github/workflows/checks.yml:58]: We currently don't have a way to skip tests that can't run on older APIC versions.

File: .github/workflows/checks.yml (hunk 5 above)
Comment (samiib) [reply .github/workflows/checks.yml]: I added the CodeCov step but it looks like we need to provide a token via a secret. However, I don't have admin access to do that myself. But it should run successfully when the CodeCov token secret is configured as an environment variable.

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**